### ISSUE TITLE
Cannot infer concrete type of templated type in intersection type expression

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#1673]
Similar to #1672, the `ConcreteParameter` type cannot be inferred at the end of this block.

```php
<?php

/** @psalm-template SomeType */
interface ParentType
{
    /** @psalm-return SomeType */
    function foo();
}

interface ChildType extends ParentType
{
}

class ConcreteParameter
{
    function sayHello() : string
    {
        return 'hi';
    }
}

/** @psalm-return ChildType&ParentType<ConcreteParameter> */
function makeConcrete() : ChildType
{
    return new class() implements ChildType {
        public function foo() {
            return new ConcreteParameter();
        }
    };
}

makeConcrete()->foo()->sayHello();
```

See https://psalm.dev/r/d50fc468f3 for examples

**Comment by muglug** [comment]
Changing to `/** @psalm-return ParentType<ConcreteParameter>&ChildType& */` fixes the call (but confuses the  `return` check

**Comment by Ocramius** [comment]
Note that I have multiple types in my case (not just 2)

**Comment by muglug** [comment]
I think this is caused by https://github.com/vimeo/psalm/issues/1674

**Comment by muglug** [comment]
Fixed by the fix to #1674
